RevenueCatUI/CustomerCenter/Data/LoadPromotionalOfferUseCase.swift

```diff
@@ -43,47 +43,103 @@ class LoadPromotionalOfferUseCase: LoadPromotionalOfferUseCaseType {
+                                            promoOfferDetails: promoOfferDetails)
+
+            let promotionalOffer = try await self.purchasesProvider.promotionalOffer(
+                forProductDiscount: discount,
+                product: subscribedProduct
+            )
+            return .success(PromotionalOfferData(
+                promotionalOffer: promotionalOffer,
+                product: subscribedProduct,
+                promoOfferDetails: promoOfferDetails
+            ))
         } catch {
             Logger.warning(Strings.error_fetching_promotional_offer(error))
             return .failure(CustomerCenterError.couldNotFindOfferForActiveProducts)
         }
     }
 
+    private func getActiveSubscription(_ customerInfo: CustomerInfo) async throws -> (String, StoreProduct) {
+        guard let productIdentifier = customerInfo.earliestExpiringAppStoreEntitlement()?.productIdentifier,
+              let subscribedProduct = await self.purchasesProvider.products([productIdentifier]).first else {
+            Logger.warning(Strings.could_not_offer_for_any_active_subscriptions)
+            throw CustomerCenterError.couldNotFindSubscriptionInformation
+        }
+        return (productIdentifier, subscribedProduct)
+    }
+
+    private func findDiscount(
+        for product: StoreProduct,
+        productIdentifier: String,
+        promoOfferDetails: CustomerCenterConfigData.HelpPath.PromotionalOffer
+    ) throws -> StoreProductDiscount {
+        let discount = if !promoOfferDetails.productMapping.isEmpty {
+            findMappedDiscount(for: product,
+                               productIdentifier: productIdentifier,
+                               promoOfferDetails: promoOfferDetails)
+        } else {
+            findLegacyDiscount(for: product, promoOfferDetails: promoOfferDetails)
+        }
+
+        guard let discount = discount else {
+            logDiscountError(productIdentifier: productIdentifier, promoOfferDetails: promoOfferDetails)
+            throw CustomerCenterError.couldNotFindSubscriptionInformation
+        }
+
+        return discount
+    }
+
+    private func findMappedDiscount(
+        for product: StoreProduct,
+        productIdentifier: String,
+        promoOfferDetails: CustomerCenterConfigData.HelpPath.PromotionalOffer
+    ) -> StoreProductDiscount? {
+        product.discounts.first { discount in
+            guard let offerIdentifier = discount.offerIdentifier else { return false }
+            return promoOfferDetails.productMapping[productIdentifier] == offerIdentifier
+        }
+    }
+
+    private func findLegacyDiscount(
+        for product: StoreProduct,
+        promoOfferDetails: CustomerCenterConfigData.HelpPath.PromotionalOffer
+    ) -> StoreProductDiscount? {
+        // Try exact match first
+        if let exactMatch = product.discounts.first(where: {
+            $0.offerIdentifier == promoOfferDetails.iosOfferId
+        }) {
```

Comment:
FFTI, but you can also do it like this
```suggestion
        if let exactMatch = product.discounts.first {
            $0.offerIdentifier == promoOfferDetails.iosOfferId
        } {
```

Then again maybe the brackets add some clarity since it ends in a `{` too

Reply:
It actually fails because of the  ending `{`